### PR TITLE
TimeKeeper

### PR DESCRIPTION
**Pull request by joelwetzel**
Adds a TimeKeeper class.  The TimeKeeper is a class that intercepts and overrides the default Date constructor.
- This allows us to control the current time in tests, without having to build anything special into the app script itself.
- Tests can advance time.
- As long as an app script uses "new Date()", or now() on the AppExecutor, to get the current time, then this class will control what time the app sees.
- Note that you must call install() before the app script is tested, and uninstall() afterwards.

### CODE REVIEW

**Comment by joelwetzel** [comment]
Ah, looks like the tests are failing in the CI build due to a timezone difference between the server and my machine.  I'll fix that up.

**Comment by joelwetzel** [comment]
Ok, I fixed it.

**Comment by joelwetzel** [comment]
Here's an example of TimeKeeper in use:  https://github.com/joelwetzel/Hubitat-Auto-Shades/blob/unitTests/tests/BehaviorTests.groovy

File: src/main/groovy/me/biocomp/hubitat_ci/util/TimeKeeper.groovy [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    TimeKeeper(Date startingDate) {
+        internalDate = startingDate
+    }
+
+    /**
+    * Install the override of the default Date constructor
+    */
+    def install() {
```
Comment (biocomp) [inline src/main/groovy/me/biocomp/hubitat_ci/util/TimeKeeper.groovy:25]: I'm not sure I'm happy with overriding the `Date` globally. Does look somewhat error-prone.
Should you instead substitute the `Date` class uses inside of the scripts (such as I did for some of them) for your own class which reads the date from a well known place?

And you can implement `AppExecutorWithEventForwarding::now()` in terms of that class too.

File: src/main/groovy/me/biocomp/hubitat_ci/util/AppExecutorWithEventForwarding.groovy [hunk 2]
```diff
@@ -15,6 +15,11 @@ abstract class AppExecutorWithEventForwarding implements AppExecutor {
 
     private HubitatAppScript script
 
+    @Override
+    long now() {
```
Comment (biocomp) [inline src/main/groovy/me/biocomp/hubitat_ci/util/AppExecutorWithEventForwarding.groovy:19]: Seems like you'll need to rename this class into some more generic like `PartialAppExecutor` (I'm happy to hear better naming suggestions :) )

File: src/main/groovy/me/biocomp/hubitat_ci/util/TimeKeeper.groovy (hunk 1 above)
Comment (biocomp) [reply src/main/groovy/me/biocomp/hubitat_ci/util/TimeKeeper.groovy:25]: Such as I do in https://github.com/biocomp/hubitat_ci/blob/9d7dc67f2c80b40267827609903d79f2b0a6a926/src/main/groovy/me/biocomp/hubitat_ci/util/SandboxClassLoader.groovy#L18C26-L18C38

File: src/main/groovy/me/biocomp/hubitat_ci/util/AppExecutorWithEventForwarding.groovy (hunk 2 above)
Comment (joelwetzel) [reply src/main/groovy/me/biocomp/hubitat_ci/util/AppExecutorWithEventForwarding.groovy:19]: Yeah, I had been thinking that as more functionality came in.  I'm working on a separate PR where I rename it to ConcreteAppExecutor.  It's a tough thing to name properly, since I don't want to implement everything.  Only the stuff that helps write more useful tests.

File: src/main/groovy/me/biocomp/hubitat_ci/util/TimeKeeper.groovy (hunk 1 above)
Comment (joelwetzel) [reply src/main/groovy/me/biocomp/hubitat_ci/util/TimeKeeper.groovy:25]: Ah, that's an interesting approach.  I will take a look.

Comment (joelwetzel) [reply src/main/groovy/me/biocomp/hubitat_ci/util/TimeKeeper.groovy:25]: I almost have it, but I may have run into a show-stopper.

Ok, so it is very common for Hubitat app and device scripts to have lines like "def now = new Date()".  That's the most common way to get the current time.  And since it can then be used in the script as a Date instance, that means that my own class that I'm substituting in needs to extend Date.

So I made a TimeKeeperDate that extends Date.  And made the TimeKeeper override its constructor, instead of overriding Date's constructor.  And I even wrote a bunch of unit tests to make sure that the existence of TimeKeeperDate, and also whether or not the TimeKeeper was installed, I want to make sure that it doesn't impact the true Date constructor.

That all worked.  So far, so good.

But then I wanted to test that SandboxClassLoader could successfully substitute in TimeKeeperDate everywhere that the script uses Date.  So I wrote an integration test for that, and here's the showstopper problem.

It looks like ClassLoader doesn't substitute classes in the "script code".  It substitutes the actual implementation of the class in memory.  

So, I added a substitution of 'java.util.Date' --> 'me.biocomp.hubitat_ci.util.TimeKeeperDate'.  And it does do that.  But then it sees that TimeKeeperDate extends java.util.Date, and of course, now it doesn't know what that is.  So it throws this exception:

```
java.lang.NoClassDefFoundError: java/util/Date
	at java.base/java.lang.Class.privateGetDeclaredMethods(Class.java:3166)
	at java.base/java.lang.Class.privateGetPublicMethods(Class.java:3191)
	at java.base/java.lang.Class.getMethods(Class.java:1904)
	at java.desktop/com.sun.beans.introspect.MethodInfo.get(MethodInfo.java:70)
	at java.desktop/com.sun.beans.introspect.ClassInfo.getMethods(ClassInfo.java:80)
	at java.desktop/java.beans.Introspector.getTargetMethodInfo(Introspector.java:1029)
	at java.desktop/java.beans.Introspector.getBeanInfo(Introspector.java:445)
	at java.desktop/java.beans.Introspector.getBeanInfo(Introspector.java:195)
	at groovy.lang.GroovyObjectSupport.getDefaultMetaClass(GroovyObjectSupport.java:59)
	at groovy.lang.GroovyObjectSupport.<init>(GroovyObjectSupport.java:32)
	at groovy.lang.Script.<init>(Script.java:42)
	at groovy.lang.Script.<init>(Script.java:39)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.Class.newInstance(Class.java:584)
	at groovy.lang.GroovyShell.parse(GroovyShell.java:555)
	at groovy.lang.GroovyShell.parse(GroovyShell.java:564)
	at me.biocomp.hubitat_ci.app.AppValidator.parseScript(AppValidator.groovy:39)
	at me.biocomp.hubitat_ci.app.HubitatAppSandbox.setupImpl(HubitatAppSandbox.groovy:74)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at me.biocomp.hubitat_ci.app.HubitatAppSandbox.run(HubitatAppSandbox.groovy:56)
```

Basically, you can't substitute a class with something that extends that class.  If we could sub it out in the script, that would work.  But actually changing the class definition isn't going to work.

Comment (joelwetzel) [reply src/main/groovy/me/biocomp/hubitat_ci/util/TimeKeeper.groovy:25]: Now, we are loading the app script from a file.  And then passing either the File or the String of the contents into GroovyShell.parse().

So, instead of switching out Date for TimeKeeperDate in the ClassLoader, I could do a text replacement before calling parse(String).  It would need to be pretty specific, because now we're dealing with text and all the many ways that developers create edge cases in it.

First stab:  find-and-replace all instances of "new Date()" with "new TimeKeeperDate()".  Enforce capitalization on the find.

2nd thought:  devs might have put extra whitespace in there, either between new and Date, or between Date and the open parenthesis, or inside the parenthesis.  A regex find-and-replace is probably better than a String find-and-replace.  Would still need a bunch of unit tests.

Comment (joelwetzel) [reply src/main/groovy/me/biocomp/hubitat_ci/util/TimeKeeper.groovy:25]: Ok, I have implemented the alternate integration:  https://github.com/biocomp/hubitat_ci/pull/53

Take a look and let me know if you prefer that.  Thanks!

**Comment by joelwetzel** [comment]
Question for ya:  I know I have some repos that I haven't touched in a few years, and they do what I need, and if someone sent me PRs for their use cases, I mean I guess it's great, but honestly I don't want to put the time in to evaluate them.  The repo already serves my needs.

Would you prefer if I not send you PRs?  I could just work in my own fork, but I wanted to offer the work back to you first.

Thanks!

File: src/main/groovy/me/biocomp/hubitat_ci/util/TimeKeeper.groovy (hunk 1 above)
Comment (biocomp) [reply src/main/groovy/me/biocomp/hubitat_ci/util/TimeKeeper.groovy:25]: I think let's go with the string replace approach, though it does seem fragile. But maybe that's good enough for most scenarios.

**Comment by biocomp** [comment]
> Would you prefer if I not send you PRs? I could just work in my own fork, but I wanted to offer the work back to you first.

I do prefer you send the PRs, thanks for doing this. This way there's two pairs of eyes watching, and maybe I can give some insights in the design. 

Sorry that I can't review all of them quickly enough.

**Comment by joelwetzel** [comment]
> > Would you prefer if I not send you PRs? I could just work in my own fork, but I wanted to offer the work back to you first.
> 
> I do prefer you send the PRs, thanks for doing this. This way there's two pairs of eyes watching, and maybe I can give some insights in the design.
> 
> Sorry that I can't review all of them quickly enough.

No worries!  I appreciate the reviews enough to be patient.  They improve the ideas.  Thanks!

File: src/main/groovy/me/biocomp/hubitat_ci/util/TimeKeeper.groovy (hunk 1 above)
Comment (joelwetzel) [reply src/main/groovy/me/biocomp/hubitat_ci/util/TimeKeeper.groovy:25]: Ok, I'll close this PR out and you should review/approve this one then:  https://github.com/biocomp/hubitat_ci/pull/53
